[GStreamer] Restrict allowed URI protocols for media playback

crmurillo opened the pull request:
#### 7e871b506eaba52c7d74519aeb0047d702101331
<pre>
[GStreamer] Restrict allowed URI protocols for media playback
<a href="https://bugs.webkit.org/show_bug.cgi?id=301092">https://bugs.webkit.org/show_bug.cgi?id=301092</a>

Reviewed by Philippe Normand.

Restrict allowed URI protocols for media playback

Add protocol validation to prevent potentially unsafe URI protocols from
being used for media playback. Only http(s), file, blob, data,
mediasourceblob, and mediastream protocols are allowed by default.

Additionally, support WEBKIT_GST_ALLOWED_URI_PROTOCOLS environment
variable to extend the list of allowed protocols via comma-separated
values.

Test: Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::isProtocolAllowed):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::load):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
(TestWebKitAPI::TEST_F(GStreamerTest, protocolValidation)):
(TestWebKitAPI::TEST_F(GStreamerTest, protocolValidationEnvironmentVariable)):

Canonical link: <a href="https://commits.webkit.org/302338@main">https://commits.webkit.org/302338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e2c5c9dfb4cd045f6ef6227a7cf48093eab86de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136184 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130672 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/935 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98057 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131748 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/767 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115383 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78666 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33496 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79464 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33990 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138644 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/875 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/851 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106598 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/928 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106406 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27092 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/721 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30251 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/53304 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/943 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64233 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/795 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/850 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/885 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -278,6 +279,35 @@ std::optional<WebCore::IntSize> getDisplaySize(WebCore::IntSize originalSize, in
+{
+    HashSet<String> allowedProtocols = { "blob"_s, "data"_s, "file"_s, "http"_s, "https"_s };
+#if ENABLE(MEDIA_SOURCE)
+        allowedProtocols.add("mediasourceblob"_s);
```
The indentation looks off here and 2 lines below.

crmurillo replied:
yes, I'll fix it

philn replied:
Thanks :)

crmurillo committed:
[GStreamer] Restrict allowed URI protocols for media playback
https://bugs.webkit.org/show_bug.cgi?id=301092

Reviewed by Philippe Normand.

Restrict allowed URI protocols for media playback

Add protocol validation to prevent potentially unsafe URI protocols from
being used for media playback. Only http(s), file, blob, data,
mediasourceblob, and mediastream protocols are allowed by default.

Additionally, support WEBKIT_GST_ALLOWED_URI_PROTOCOLS environment
variable to extend the list of allowed protocols via comma-separated
values.

Test: Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::isProtocolAllowed):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::load):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
(TestWebKitAPI::TEST_F(GStreamerTest, protocolValidation)):
(TestWebKitAPI::TEST_F(GStreamerTest, protocolValidationEnvironmentVariable)):

Canonical link: https://commits.webkit.org/302338@main
